rust/sdk-processor/src/steps/common/parquet_buffer_step.rs

```diff
@@ -0,0 +1,387 @@
+use crate::{
+    parquet_processors::{ParquetTypeEnum, ParquetTypeStructs},
+    steps::common::gcs_uploader::Uploadable,
+};
+use anyhow::Result;
+use aptos_indexer_processor_sdk::{
+    traits::{
+        pollable_async_step::PollableAsyncRunType, NamedStep, PollableAsyncStep, Processable,
+    },
+    types::transaction_context::{TransactionContext, TransactionMetadata},
+    utils::errors::ProcessorError,
+};
+use async_trait::async_trait;
+use mockall::mock;
+use std::{collections::HashMap, time::Duration};
+use tracing::debug;
+
+/// `ParquetBuffer` is a struct that holds `ParquetTypeStructs` data
+/// and tracks the buffer size in bytes, along with metadata about the data in the buffer.
+struct ParquetBuffer {
+    pub buffer: ParquetTypeStructs,
+    pub buffer_size_bytes: usize,
+    current_batch_metadata: Option<TransactionMetadata>,
+}
+
+impl ParquetBuffer {
+    fn new(parquet_type: &ParquetTypeEnum) -> Self {
+        Self {
+            buffer: ParquetTypeStructs::default_for_type(parquet_type),
+            buffer_size_bytes: 0,
+            current_batch_metadata: None,
+        }
+    }
+
+    /// Updates the metadata of the internal buffer with the latest information from `cur_batch_metadata`.
+    /// This is used to track the end version and timestamp of the batch data in the buffer.
+    pub fn update_current_batch_metadata(&mut self, cur_batch_metadata: &TransactionMetadata) {
+        if let Some(buffer_metadata) = &mut self.current_batch_metadata {
+            // Update metadata fields with the current batch's end information
+            buffer_metadata.end_version = cur_batch_metadata.end_version;
+            buffer_metadata.total_size_in_bytes = cur_batch_metadata.total_size_in_bytes;
+            buffer_metadata.end_transaction_timestamp =
+                cur_batch_metadata.end_transaction_timestamp.clone();
+        } else {
+            // Initialize the metadata with the current batch's start information
+            self.current_batch_metadata = Some(cur_batch_metadata.clone());
+        }
+    }
+}
+
+/// `ParquetBufferStep` is a step that accumulates data in buffers until they reach a specified size limit.
+///
+/// It then uploads the buffered data to Google Cloud Storage (GCS) through an uploader.
+/// This step is typically used to manage large data volumes efficiently by buffering and uploading
+/// only when necessary.
+///
+///
+/// # Type Parameters
+/// - `U`: A type that implements the `Uploadable` trait, providing the uploading functionality.
+pub struct ParquetBufferStep<U>
+where
+    U: Uploadable + Send + 'static + Sync,
+{
+    internal_buffers: HashMap<ParquetTypeEnum, ParquetBuffer>,
+    pub poll_interval: Duration,
+    pub buffer_uploader: U,
```

Comment:
Does `buffer_uploader` need to be a trait? It looks like the type will always be `GCSUploader`.